### PR TITLE
Adding compatibility for GHC 9.6, GHC 9.8, and GHC 9.10

### PR DESCRIPTION
**Pull request by recursion-ninja**
This PR adds compatibility support for the modern compiler versions of GHC; i.e. GHC 9.6, GHC 9.8, and GHC 9.10. Given the immense value of the Project-M36 database system, it would be a shame to have it succumb to "bit-rot" at the Haskell ecosystem moves forward. 

Additionally, the PR cleans up compiler warnings emitted by the newer versions of GHC. Most of the warnings were regarding calls to `head` and `tail` generated from `-W-x-partial`.

### CODE REVIEW

File: examples/Plantfarm.hs [hunk 1]
```diff
@@ -209,14 +213,14 @@ main = do
       S.json ps
 
 handleWebError :: Either Err b -> S.ActionM b
-handleWebError (Left e) = S.raise (TL.pack $ "An error occurred:\n" <> show e)
+handleWebError (Left e) = error $ "An error occurred:\n" <> show e
```
Comment (ysangkok) [inline examples/Plantfarm.hs:216]: I suppose you're replacing this because `raise` is deprecated. Wouldn't it be better to migrate `throw` which isn't deprecated? Maybe it's better to do this in a separate PR and leave `raise` alone for now, since it is still in the latest scotty. You can disable the warning for the module using OPTIONS_GHC.

File: examples/SimpleClient.hs [hunk 2]
```diff
@@ -24,7 +26,7 @@ main = do
           mErr1 <- executeDatabaseContextExpr sessionId conn (Define "person" (map NakedAttributeExpr attrList))
           print mErr1
           --5. add a tuple to the relation referenced by the relation variable
-          let (Right tupSet) = mkTupleSetFromList attrs [[TextAtom "Bob", IntAtom 45]]
+          let tupSet = fromRight undefined $ mkTupleSetFromList attrs [[TextAtom "Bob", IntAtom 45]]
```
Comment (ysangkok) [inline examples/SimpleClient.hs:29]: This doesn't seem like an improvement. Previously, the error would not that it is a pattern match failure. Now you'd get whatever `undefined` emits...

File: examples/Plantfarm.hs (hunk 1 above)
Comment (recursion-ninja) [reply examples/Plantfarm.hs]: That's a fair assessment. The motivation *was* simply to suppress the "deprecation" warning. I do concur that switching from `raise` to `throw`should be a separate PR. With this in mind I have reverted this change.

File: examples/SimpleClient.hs (hunk 2 above)
Comment (recursion-ninja) [reply examples/SimpleClient.hs]: The motivation here was also to suppress the "incomplete pattern match" warning. The solution I inserted is admittedly dubious and less clear than the original code, made in haste to quickly suppress the error. Given your suggestion to use a module specific `ghc` option flag in the `examples/Plantfarm.hs` module, I have reverted these changes and instead suppressed the warning with a `OPTIONS_GHC` pragma.

**Comment by recursion-ninja** [comment]
@ysangkok Let me know if there's any other comments on the PR.

**Comment by ysangkok** [comment]
I don't have more comments, but I am just drive-by reviewing, I don't have a commit bit or anything.

**Review by agentm — COMMENTED** [review]
Thanks so much for doing the heavy lifting to get the new compilers supported! I'll be testing this over the next few days.

This project has been struggling with developing a policy for tracking GHC and library upgrades. For example, for some years, Project:M36 was held back by the stale and unmaintained distributed-process library which we replaced with curryer-rpc. More recently, winery and stm-hamt haven't been updated and sometimes cannot be updated even after loosening version constraints due to compiler errors. In the general case, we then have the following options:

1. push fixes upstream to the dependent repository while gauging if the module is still maintained (often waiting for new releases)
1. rely on head.hackage, perhaps temporarily (new in this PR)
1. replace or drop the dependency (example: distributed-process -> curryer-rpc)

This PR points at the head.hackage project which is OK, but is it a good long-term solution?

In general, I would like to drop modules that hold us back. I like winery, but we're not married to it and it includes several dependencies which need updates for new GHCs such as barbies-th.

Do you have any feedback on identifying and replacing any of the modules which hold us back? I would very much like Project:M36 to stay up-to-date with GHC.

In any case, I'll work on getting curryer-rpc working on GHC 9.6 - 9.10 so that we can remove it from this cabal.project.

File: src/lib/ProjectM36/DataTypes/SQL/Null.hs [hunk 3]
```diff
@@ -298,9 +298,18 @@ sqlNullableIntegerToMaybe _ = Nothing
 sqlEqualsTypes :: Atom -> Atom -> Bool
 sqlEqualsTypes a b = underlyingType a == underlyingType b
   where
-    underlyingType (ConstructedAtom "SQLNull" (ConstructedAtomType "SQLNullable" typmap) []) | M.size typmap == 1 = snd (head (M.assocs typmap))
-    underlyingType (ConstructedAtom "SQLJust" (ConstructedAtomType "SQLNullable" typmap) _args) | M.size typmap == 1 = snd (head (M.assocs typmap))
-    underlyingType atom = atomTypeForAtom atom
+    underlyingType atom =
```
[the remark this answers is not in the document]
Comment (recursion-ninja) [reply src/lib/ProjectM36/DataTypes/SQL/Null.hs:301]: This change suppressed a GHC warning with newer compiler versions. I spent a while reasoning that the semantics were the same.

**Comment by recursion-ninja** [comment]
> This PR points at the `head.hackage` project which is OK, but is it a good long-term solution?

I would say that it's *not* a long term solution and upstream dependencies *should ideally* be updated to be compatible and on `hackage`. However, using `head.hackage` is a useful stepping stone.

> Do you have any feedback on identifying and replacing any of the modules which hold us back?

I didn't really do any dependency analysis with the perspective of dropping "problem" dependencies. I can take another look and create a GitHub issue with a possible migration plan.

**Comment by agentm** [comment]
Thanks! In the meantime, I opened PRs in:

* [barbies-th](https://github.com/fumieval/barbies-th/pull/14)
* [fast-builder](https://github.com/takano-akio/fast-builder/pull/16)

Once those are deployed to hackage, we should be able to update winery with the new dependency versions.
